src/DynamoCore/Models/WorkspaceModel.cs

```diff
@@ -687,6 +684,55 @@ public AnnotationModel AddAnnotation(string text, Guid id)
         }
 
         /// <summary>
+        /// Sets the event on Annotaiton model.
+        /// When a node is removed from a group, this event will be
+        /// used to include that node again in that group (UNDO operation)
+        /// </summary>
+        /// <param name="model">The model.</param>
+        internal void SetGetModelsForGrouping(AnnotationModel model)
+        {
+            model.GetModelBaseEvent += annotationModel_GetModelBaseEvent;
+        }
+
+        /// <summary>
+        /// Get the model from Workspace
+        /// </summary>
+        /// <param name="modelGuid">The model unique identifier.</param>
+        /// <returns></returns>
+        private ModelBase annotationModel_GetModelBaseEvent(Guid modelGuid)
+        {
+            ModelBase model = null;
+            model = this.Nodes.FirstOrDefault(x => x.GUID == modelGuid);
+            if (model == null) //Check if GUID is a Note instead.
+            {
+                model = this.Notes.FirstOrDefault(x => x.GUID == modelGuid);
+            }
+
+            return model;
+        }
+
+        /// <summary>
+        /// Checks if model exists in same group.
+        /// </summary>
+        /// <param name="selectNodes">The select nodes.</param>
+        /// <param name="selectNotes">The select notes.</param>
+        /// <returns>true if the models are already in the same group</returns>
+        private bool CheckIfModelExistsInSameGroup(IEnumerable<NodeModel> selectNodes, IEnumerable<NoteModel> selectNotes)
+        {
+            var selectedModels = selectNodes.Concat(selectNotes.Cast<ModelBase>()).ToList();
+            bool nodesInSameGroup = false;
+            foreach (var group in this.Annotations)
+            {
+                var groupModels = group.SelectedModels;
+                nodesInSameGroup = selectedModels.Except(groupModels).ToList().Count <= 0;
```

Comment:
Please look into what happens when you say `ToList` on an `IEnumerable`.  It causes the creation of a new list.  This only makes sense if you might be enumerating the list multiple times and you know it won't change.  Otherwise, you can just use `Count()`.  This is a common pitfall.